File: clients/go-sdk/client/roles.go
package client

import (
	"context"
	"fmt"
)

// ─── Request DTOs ────────────────────────────────────────────────────

// CreateRoleRequest — POST /api/roles.
type CreateRoleRequest struct {
	ApplicationCode string   `json:"applicationCode"`
	RoleName        string   `json:"roleName"`
	DisplayName     string   `json:"displayName"`
	Description     string   `json:"description,omitempty"`
	Permissions     []string `json:"permissions,omitempty"`
	ClientManaged   bool     `json:"clientManaged,omitempty"`
}

// UpdateRoleRequest — PUT /api/roles/{name}.
type UpdateRoleRequest struct {
	DisplayName   *string `json:"displayName,omitempty"`
	Description   *string `json:"description,omitempty"`
	ClientManaged *bool   `json:"clientManaged,omitempty"`
}

// GrantPermissionRequest — body for POST /api/roles/{name}/permissions.
type GrantPermissionRequest struct {
	Permission string `json:"permission"`
}

// ─── Response DTOs ───────────────────────────────────────────────────

// RoleResponse is the platform's role aggregate.
type RoleResponse struct {
	ID              string   `json:"id"`
	Name            string   `json:"name"`
	ShortName       string   `json:"shortName"`
	DisplayName     string   `json:"displayName"`
	Description     string   `json:"description,omitempty"`
	ApplicationCode string   `json:"applicationCode"`
	Permissions     []string `json:"permissions,omitempty"`
	Source          string   `json:"source"`
	ClientManaged   bool     `json:"clientManaged,omitempty"`
	CreatedAt       string   `json:"createdAt"`
	UpdatedAt       string   `json:"updatedAt"`
}

// RoleListResponse — GET /api/roles.
type RoleListResponse struct {
	Roles []RoleResponse `json:"roles"`
	Total uint64         `json:"total,omitempty"`
}

// ─── Sync ────────────────────────────────────────────────────────────

// SyncRoleItem — one entry in the per-app sync payload.
type SyncRoleItem struct {
	Name          string   `json:"name"`
	DisplayName   string   `json:"displayName,omitempty"`
	Description   string   `json:"description,omitempty"`
	Permissions   []string `json:"permissions,omitempty"`
	ClientManaged bool     `json:"clientManaged,omitempty"`
}

// SyncRolesRequest — body for the per-app sync endpoint.
type SyncRolesRequest struct {
	Roles []SyncRoleItem `json:"roles"`
}

// ─── Resource ────────────────────────────────────────────────────────

// RolesResource — /api/roles/*.
type RolesResource struct {
	c *FlowCatalystClient
}

// List — GET /api/roles.
func (r *RolesResource) List(ctx context.Context) (*RoleListResponse, error) {
	var out RoleListResponse
	if err := r.c.Get(ctx, "/api/roles", &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Get — GET /api/roles/{name}.
func (r *RolesResource) Get(ctx context.Context, name string) (*RoleResponse, error) {
	var out RoleResponse
	if err := r.c.Get(ctx, "/api/roles/"+name, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// GetByCode — GET /api/roles/by-code/{code} (`application:role-name`).
func (r *RolesResource) GetByCode(ctx context.Context, code string) (*RoleResponse, error) {
	var out RoleResponse
	if err := r.c.Get(ctx, "/api/roles/by-code/"+code, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Create — POST /api/roles. Returns { id } only; call Get(name) for
// the full record.
func (r *RolesResource) Create(ctx context.Context, req *CreateRoleRequest) (*CreatedResponse, error) {
	var out CreatedResponse
	if err := r.c.Post(ctx, "/api/roles", req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Update — PUT /api/roles/{name}. Platform responds with 204.
func (r *RolesResource) Update(ctx context.Context, name string, req *UpdateRoleRequest) error {
	return r.c.Put(ctx, "/api/roles/"+name, req, nil)
}

// Delete — DELETE /api/roles/{name}.
func (r *RolesResource) Delete(ctx context.Context, name string) error {
	return r.c.Delete(ctx, "/api/roles/"+name, nil)
}

// ListForApplication — GET /api/roles/by-application/{applicationId}.
func (r *RolesResource) ListForApplication(ctx context.Context, applicationID string) (*RoleListResponse, error) {
	var out RoleListResponse
	if err := r.c.Get(ctx, "/api/roles/by-application/"+applicationID, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// GrantPermission — POST /api/roles/{name}/permissions. Returns the updated role.
func (r *RolesResource) GrantPermission(ctx context.Context, roleName, permission string) (*RoleResponse, error) {
	var out RoleResponse
	if err := r.c.Post(ctx, "/api/roles/"+roleName+"/permissions",
		&GrantPermissionRequest{Permission: permission}, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// RevokePermission — DELETE /api/roles/{name}/permissions/{permission}. Returns the updated role.
func (r *RolesResource) RevokePermission(ctx context.Context, roleName, permission string) (*RoleResponse, error) {
	var out RoleResponse
	if err := r.c.Delete(ctx, fmt.Sprintf("/api/roles/%s/permissions/%s", roleName, permission), &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Sync — POST /api/applications/{appCode}/roles/sync.
func (r *RolesResource) Sync(ctx context.Context, appCode string, req *SyncRolesRequest, removeUnlisted bool) (*SyncResult, error) {
	q := ""
	if removeUnlisted {
		q = "?removeUnlisted=true"
	}
	var out SyncResult
	if err := r.c.Post(ctx, fmt.Sprintf("/api/applications/%s/roles/sync%s", appCode, q), req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

File: crates/fc-sdk/src/client/roles.rs
//! Role management operations.
//!
//! Permission-catalogue queries live on a sibling [`crate::client::Permissions`]
//! accessor.

use super::applications::CreatedResponse;
use super::{ClientError, FlowCatalystClient};
use serde::{Deserialize, Serialize};

/// Paginated list of roles — `GET /api/roles`.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RoleListResponse {
    pub roles: Vec<RoleResponse>,
    #[serde(default)]
    pub total: u64,
}

/// Request to create a role.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct CreateRoleRequest {
    /// Application code this role belongs to
    pub application_code: String,
    /// Role name (will be combined with app code to form code)
    pub role_name: String,
    /// Display name
    pub display_name: String,
    /// Description
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    /// Initial permissions
    #[serde(default)]
    pub permissions: Vec<String>,
    /// Whether clients can manage this role
    #[serde(default)]
    pub client_managed: bool,
}

/// Request to update a role.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct UpdateRoleRequest {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub display_name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub client_managed: Option<bool>,
}

/// Request body for `grant_permission` — `{ "permission": "..." }`.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GrantPermissionRequest {
    pub permission: String,
}

/// Role response from the platform API.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RoleResponse {
    pub id: String,
    pub name: String,
    pub short_name: String,
    pub display_name: String,
    #[serde(default)]
    pub description: Option<String>,
    pub application_code: String,
    #[serde(default)]
    pub permissions: Vec<String>,
    pub source: String,
    #[serde(default)]
    pub client_managed: bool,
    pub created_at: String,
    pub updated_at: String,
}

/// Request body for the per-resource sync endpoint.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SyncRolesRequest {
    pub roles: Vec<SyncRoleItem>,
}

/// A role item for sync.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SyncRoleItem {
    pub name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub display_name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(default)]
    pub permissions: Vec<String>,
    #[serde(default)]
    pub client_managed: bool,
}

/// Roles resource accessor — created via [`FlowCatalystClient::roles`].
pub struct Roles<'a> {
    pub(crate) client: &'a FlowCatalystClient,
}

impl Roles<'_> {
    /// List all roles.
    pub async fn list(&self) -> Result<RoleListResponse, ClientError> {
        self.client.get("/api/roles").await
    }

    /// Get a role by name.
    pub async fn get(&self, name: &str) -> Result<RoleResponse, ClientError> {
        self.client.get(&format!("/api/roles/{}", name)).await
    }

    /// Get a role by code (`application:role-name`).
    pub async fn get_by_code(&self, code: &str) -> Result<RoleResponse, ClientError> {
        self.client
            .get(&format!("/api/roles/by-code/{}", code))
            .await
    }

    /// Create a new role.
    ///
    /// Returns `{ id }` only. Call `get(name)` if you need the full record.
    pub async fn create(
        &self,
        req: &CreateRoleRequest,
    ) -> Result<CreatedResponse, ClientError> {
        self.client.post("/api/roles", req).await
    }

    /// Update an existing role by name. The platform responds with 204.
    pub async fn update(
        &self,
        name: &str,
        req: &UpdateRoleRequest,
    ) -> Result<(), ClientError> {
        let _: serde_json::Value = self
            .client
            .put(&format!("/api/roles/{}", name), req)
            .await?;
        Ok(())
    }

    /// Delete a role by name.
    pub async fn delete(&self, name: &str) -> Result<(), ClientError> {
        self.client
            .delete_req(&format!("/api/roles/{}", name))
            .await
    }

    /// List roles scoped to an application.
    pub async fn list_for_application(
        &self,
        application_id: &str,
    ) -> Result<RoleListResponse, ClientError> {
        self.client
            .get(&format!("/api/roles/by-application/{}", application_id))
            .await
    }

    /// Grant a permission to a role. Returns the updated role.
    pub async fn grant_permission(
        &self,
        role_name: &str,
        permission: &str,
    ) -> Result<RoleResponse, ClientError> {
        let body = GrantPermissionRequest {
            permission: permission.to_string(),
        };
        self.client
            .post(&format!("/api/roles/{}/permissions", role_name), &body)
            .await
    }

    /// Revoke a permission from a role. Returns the updated role.
    pub async fn revoke_permission(
        &self,
        role_name: &str,
        permission: &str,
    ) -> Result<RoleResponse, ClientError> {
        self.client
            .delete_with_response(&format!(
                "/api/roles/{}/permissions/{}",
                role_name, permission
            ))
            .await
    }

    /// Sync roles for an application — declarative reconciliation against
    /// `POST /api/applications/{appCode}/roles/sync`.
    pub async fn sync(
        &self,
        app_code: &str,
        req: &SyncRolesRequest,
        remove_unlisted: bool,
    ) -> Result<crate::client::SyncResult, ClientError> {
        let query = if remove_unlisted {
            "?removeUnlisted=true"
        } else {
            ""
        };
        self.client
            .post(
                &format!("/api/applications/{}/roles/sync{}", app_code, query),
                req,
            )
            .await
    }
}
